examples/example_fec_caching.py

```diff
@@ -0,0 +1,98 @@
+"""
+Example demonstrating the use of the caching decorator with functional equivalance checking
+===========================================================================================
+
+Caches the results of fitness evaluations in a pickle file
+('example_fec_caching_cache.pkl'). To illustrate its practical use,
+compare the runtime of this script when you first call it vs. the
+second time and when you comment out the decorator on
+`inner_objective`.
+
+"""
+
+import numpy as np
+import time
+
+import cgp
+
+# %%
+# We define the target function for this example.
+
+
+def f_target(x):
+    return x ** 2 + x + 1.0
+
+
+# %%
+# We then define the objective function for the evolutionary
+# algorithm: It consists of an inner objective which we wrap with the
+# caching decorator. This decorator specifies a pickle file that will be used for
+# caching results of fitness evaluations. The inner objective is then used by the objective
+# function to compute (or retrieve from cache) the fitness of the individual.
+
+
+@cgp.utils.disk_cache("example_fec_caching_cache.pkl", use_fec=True, fec_seed=12345)
+def inner_objective(ind):
+    """The caching decorator uses the return values generated from
+    providing random inputs to ind.to_numpy() to identify functionally
+    indentical individuals and avoid reevaluating them. Note that
+    caching only makes sense for deterministic objective functions, as
+    it assumes that identical phenotypes will always return the same
+    fitness values.
+
+    """
+    f = ind.to_func()
+
+    loss = []
+    for x_0 in np.linspace(-2.0, 2.0, 100):
+        y = f([x_0])
+        loss.append((f_target(x_0) - y) ** 2)
+
+    time.sleep(0.25)  # emulate long fitness evaluation
```

Comment:
Out of curiosity: What is this sleep timer for?

Reply:
this is just to make the objective take a while. if this is not clear  from the comment, maybe it should adapt it?

Reply:
That I understood, what I am wondering is why the objective should take a while?

Reply:
ahaha, sorry! :D sleep is just to make the difference between using the decorator and not using the decorator obvious. without the sleep here would be no big difference as running the objective once takes about as much time as loading the result from disk. caching is only useful for long running fitness evaluations and the sleep tries to mimic this.